### PR TITLE
Handle selecting a reflection table with experiment identifiers

### PR DESCRIPTION
**Pull request by jbeilstenedmands**
This short PR introduces some new functionality for handling reflection table selections when experiment identifiers are set, to form the basis for future work.

As suggested by @dagewa in #902, the experiment identifiers should be updated when making a selection. So for the case of `new_refl = refl.select(selection_array)`, e.g. `new_refl = refl.select(refl["id"] == 0)`,  if `refl` contains an id column and identifiers are set, then the identifiers map is now set in `new_refl`for all ids in `new_refl["id"]`. This differs from the current behaviour, where all of the identifiers are copied across to the new table, which could cause issues later on.

Also implemented are two new select methods, for passing an Experiment or ExperimentList to reflections.select(), to allow
```
for exp in experiments:
    refl = reflections.select(exp)
```
or
```
for elist in experimentlists:
    refl = reflections.select(elist)
```
as opposed to the current motif
```
for i, exp in enumerate(experiments):
     refl = reflections.select(refl["id"] == i)
```
for a reflection table `reflections` containing data from multiple experiments.

### CODE REVIEW

File: test/array_family/test_identifiers_handling.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    If no identifiers are set, then nothing should happen."""
+
+    # Test when identifiers & id column set.
+    refl = flex.reflection_table()
+    refl["id"] = flex.int([0, 0, 0, 1, 1, 1, 2, 2, 2])
+    refl.experiment_identifiers()[0] = "0"
+    refl.experiment_identifiers()[1] = "1"
+    refl.experiment_identifiers()[2] = "2"
+
+    new_refl = refl.select(refl["id"] == 0)
```
Comment (dagewa) [inline test/array_family/test_identifiers_handling.py]: Might be worth testing using `refl["id"] == 1` instead as a demonstration that identifiers are _not_ reset to start from 0 in the new table

File: test/array_family/test_identifiers_handling.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    refl.experiment_identifiers()[2] = "2"
+    experiments = ExperimentList(
+        [
+            Experiment(identifier="0"),
+            Experiment(identifier="2"),
+            Experiment(identifier="3"),
+        ]
+    )
+
+    sel_refl = refl.select(experiments)
```
Comment (dagewa) [inline test/array_family/test_identifiers_handling.py]: This line prints
```
0
2
3
```

File: array_family/boost_python/flex_table_suite.h [hunk 3]
```diff
@@ -668,6 +710,80 @@ namespace dials { namespace af { namespace boost_python { namespace flex_table_s
...
+
+  template <typename T>
+  T select_using_experiments(T &self, dxtbx::model::ExperimentList expts){
+    typedef typename T::experiment_map_type::const_iterator const_iterator;
+    typedef dxtbx::model::ExperimentList::shared_type::const_iterator expt_const_iterator;
+    T result;
+    for (expt_const_iterator expt = expts.begin(); expt != expts.end(); ++expt) {
+
+      std::string identifier = expt->get_identifier();
+      std::cout << identifier << std::endl;
```
Comment (dagewa) [inline array_family/boost_python/flex_table_suite.h]: Found the source of the print in the tests. This should be removed

**Comment by jbeilstenedmands** [comment]
The failing test has highlighted an important consideration. My code assumes that the id values are ints, but you could have size_t or others i guess. Perhaps I need some templating here...

**Pull request merged**